### PR TITLE
Add web routes for market, location pages.  Add Route::getLocation macro #85

### PR DESCRIPTION
**Pull request by pdbreen**
* [x] requires updated tipoff/support containing `view/layout.blade.php` prior to merge

---

Adds routes
- {prefix}/{market->slug}/{location->slug} => view(locations::location)
- {prefix}/{market->slug} => view(locations::market)

Fixes `ResolveLocation` middleware to work properly in package (must be applied after SubstituteBindings).  Exceptions from failure to properly resolve route due to market or location ambiguity get rendered as a view(locations::location_select) with either a full or market constrained list of locations.

Adds `MarketResolver` to support single market scenario.  Updates `LocationResolver` and `MarketResolver` to store resolved market and location in application container.  Each resolver has a static helper method to fetch Market or Location from the container or null if no context exists.

Adds `Route::getLocation($routeName, $uri, $action)` macro that should be used to properly register all route variations of a location dependent page.  The action controller can have a `Location $location` parameter (and `Market $market` - but market can always be obtained from location).  

Updates `LocationRouter::build($routeName, $location)` to build the appropriate type of URL based on market/location ambiguity that exists.  The `$location` parameter can be omitted to use the current location in context.

### CODE REVIEW

File: tests/Unit/Http/Middleware/ResolveLocationTest.php [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+{
+    use DatabaseTransactions;
+
+    /** @test */
+    public function single_market_single_location()
+    {
+        $this->logToStderr();
+        Route::getLocation('test', 'test', function () {
+            return LocationRouter::build('test');
+        });
```
Comment (pdbreen) [inline tests/Unit/Http/Middleware/ResolveLocationTest.php:24]: @drewroberts this is an example of how other packages should register location dependent routes.  Other than passing route name as first parameter, syntax is same as `Route::get(...)` - this test version is using an inline function but all variations of controller definition remain supported as well.

edit:  these would be in the other packages `routes/web.php` file.

**Comment by pdbreen** [comment]
You can see from the blade files, I spent no effort on their content - I focused only on getting the framework in place to support their output.

File: src/ViewComposers/LocationSelectComposer.php [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    {
+        $market = MarketResolver::market();
+        $locations = $market ? $market->locations : Location::all();
+
+        $view->with([
+            'market' => $market,
+            'locations' => $locations,
+        ]);
+    }
+}
```
Comment (pdbreen) [inline src/ViewComposers/LocationSelectComposer.php:22]: Using a view composer makes it easy to render the `locations::location_select` view from anywhere - like from an exception caused by ambiguous location context.

File: tests/Unit/Http/Middleware/ResolveLocationTest.php (hunk 1 above)
Comment (drewroberts) [reply tests/Unit/Http/Middleware/ResolveLocationTest.php:24]: This is great! Thank you!

File: src/ViewComposers/LocationSelectComposer.php (hunk 2 above)
Comment (drewroberts) [reply src/ViewComposers/LocationSelectComposer.php:22]: I like this!

**Review by drewroberts — APPROVED** [review]
Thank you, Patrick! 🥳

**Pull request merged**